components/builder-depot/src/server.rs

```diff
@@ -1779,12 +1691,50 @@ fn show_package(req: &mut Request) -> IronResult<Response> {
                 session_id,
                 &ident.get_origin(),
             ));
-            request.set_ident(ident);
+            request.set_ident(ident.clone());
+            let u = helpers::extract_query_value("upstream", req);
+
+            // If the "upstream" query parameter is not passed, then we can assume it's a normal
+            // API request and we should check for an update in the background.
+            if u.is_none() {
+                refresh_package_from_upstream_depot(req, ident.clone(), Some(channel.clone()));
+            }
 
             match route_message::<OriginChannelPackageLatestGet, OriginPackageIdent>(req, &request)
             {
                 Ok(id) => ident = id.into(),
-                Err(err) => return Ok(render_net_error(&err)),
+                Err(err) => {
+                    let code = err.get_code();
+
+                    // If the package is not found and it's the hab CLI making the request and
+                    // upstream=t is passed in the query string, we can assume someone is trying to
+                    // hab pkg install this particular package, and we want to synchronously check
+                    // the upstream depot before returning a 404 (assuming an upstream depot is
```

Comment:
We can, but my main thought here was the UX.  If I run `hab pkg install` and the package I want to install doesn't exist in the depot that I'm pointed at, what should happen?

1) We return the normal error that we normally do when a package doesn't exist, which says "package doesn't exist" or something similar, and then we check the upstream in the background.  That's great from an efficiency perspective, but it's a bummer from a user's perspective.  Now they might think they're stuck, unable to get the package they need.  If they try it again, of course, it will likely succeed this time, assuming the upstream has it and we've downloaded it in the background and enough time has elapsed between their first request and their second, but I'm guessing many (most?) users wouldn't think to try again.

2) We check and download it synchronously within the request.  This felt like the best UX to me because you get immediate results.  The biggest downside here is the person is left waiting a bit longer than normal, because the depot has to download the package from the upstream and process it before it can send it along to the user.  But this synchronous downloading of the upstream package only happens when we've detected that the user is running `hab pkg install`, so they're likely ok with waiting a little bit to get the package they want.

@ryankeairns do you have any thoughts on the best UX here?

Reply:
In either case, erring on the side of clarity would be prudent. For example, in the first scenario, could you provide words to inform them to try again? "That package doesn't exist/could not be found. If you are certain it does exist in the target repo, then you might try again."

In the second case, a short message like "...we're trying to find it in [here]" or "...we're downloading/processing from the upstream depot..." would inform them why it may feel slow.

Reply:
Returning "Not Found" to the user seems perfectly valid (and correct) - the package literally isn't in the local depot. We can tweak the cli UX, but it might be better to simply advise admins to bootstrap core packages using the on-prem tool we already have for now - that minimizes (maybe even eliminates) the UX issue.  The "not found" case is a very small subset of the staleness case, I don't think we should worry about "not found".  The entire strategy for now is to let the update happen in the background, and we should stick with that strategy.  This will all have a very nice fix once we have notifications from the upstream - we'll refresh (in the background) on notification from the upstream, thus greatly minimizing the staleness window.

Reply:
Ok, I'll make that change.